### ISSUE TITLE
Test failure on PPC with V8 5.9

### ISSUE THREAD

**Issue by targos** [nodejs/v8#8]

**Comment by targos** [comment]
https://ci.nodejs.org/job/node-test-commit-aix/4543/nodes=aix61-ppc64/console
https://ci.nodejs.org/job/node-test-commit-plinux/7676/nodes=ppcle-ubuntu1404/console

Error:
```
not ok 568 parallel/test-http-pipeline-regr-3332
  ---
  duration_ms: 0.608
  severity: fail
  stack: |-
    _stream_writable.js:314
      var len = state.objectMode ? 1 : chunk.length;
                                            ^
    
    TypeError: Cannot read property 'length' of undefined
        at writeOrBuffer (_stream_writable.js:314:41)
        at Socket.Writable.write (_stream_writable.js:257:11)
        at Socket.write (net.js:694:40)
        at ServerResponse._flushOutput (_http_outgoing.js:863:18)
        at ServerResponse._flush (_http_outgoing.js:840:16)
        at ServerResponse.assignSocket (_http_server.js:163:8)
        at resOnFinish (_http_server.js:520:9)
        at emitNone (events.js:112:20)
        at ServerResponse.emit (events.js:209:7)
        at onFinish (_http_outgoing.js:732:10)
```

**Comment by targos** [comment]
still happening with 5.9.82
https://ci.nodejs.org/job/node-test-commit-plinux/7724/nodes=ppcbe-ubuntu1404/console
https://ci.nodejs.org/job/node-test-commit-aix/4593/nodes=aix61-ppc64/console
https://ci.nodejs.org/job/node-test-commit-linuxone/4682/nodes=rhel72-s390x/console

@nodejs/platform-aix @nodejs/platform-ppc

**Comment by gibfahn** [comment]
cc/ @joransiu @jbarz @jbajwa

**Comment by joransiu** [comment]
From link above, it also fails the same way on s390x.

**Comment by jbajwa** [comment]
I'm looking into it now.

**Comment by jbajwa** [comment]
looks like it started failing after TF was turned on by default ([this](https://github.com/v8/v8/commit/0f716acadaed1d9e194593543dbe1340d600d6fc) commit). If I set the turbofan flag off, I don't see this issue. 
There are no known issues with TF on ppc/s390 with standalone current V8. Will investigate further.

**Comment by targos** [comment]
It is still failing with 5.9.160: https://ci.nodejs.org/job/node-test-commit/8869/

**Comment by jbajwa** [comment]
Update on the issue:
Looks like the issue is in the deoptimizer, the frames are not setup correctly when it bails out from optimized code. Deoptimizing of these js functions causes the failure: `_flushOutput` , `Writable.write` , `writeOrBuffer`.
If I force turbofan to compile code for any of these functions the testcase passes, for eg:
`out/Release/node --turbo-filter=_flushOutput test/parallel/test-http-pipeline-regr-3332.js`
`out/Release/node --turbo-filter=Writable.write test/parallel/test-http-pipeline-regr-3332.js`
`out/Release/node --turbo-filter=writeOrBuffer test/parallel/test-http-pipeline-regr-3332.js`
And, passing `--always-opt` would pass.
I can also confirm this by looking at the `--trace-deopt`, and whenever the deopt code is used for any of the above functions, the testcases fails.
Will spend sometime on it this week, hopefully will have a fix by end of the week.

**Comment by john-yan** [comment]
Hello @targos , could you please point me to the branch that causes the failure? because all the links above are outdated.

**Comment by targos** [comment]
@john-yan it's the [canary](https://github.com/nodejs/node/tree/canary) branch. I updated it today with V8 6.0.61.
Here is a new test run: https://ci.nodejs.org/job/node-test-commit/9383/

**Comment by jbajwa** [comment]
Hi,
I've just committed the [fix](https://codereview.chromium.org/2839343003) to V8 master.
Although, the current canary branch at V8 6.0.92 contains an incomplete fix I added yesterday which fixed this issue but caused another testcase to fail, I've fixed it and relanded. Should be clean now.

**Comment by mhdawson** [comment]
@targos have you validated that this is resolved ?

**Comment by targos** [comment]
It is resolved for 6.0. We can either close this issue or keep it open until we update master to 5.9, because we'll have to backport the fix.

**Comment by jbajwa** [comment]
@targos 
Hi, I forgot to mention but I did backport the fix to [5.9](https://chromium-review.googlesource.com/c/488781/)

**Comment by targos** [comment]
@jbajwa Oh that's perfect, thank you! Closing then...
